packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.tsx

```diff
@@ -1,102 +1,141 @@
-      anySize.push({ label: t('label.any'), value: -1 });
-    }
-    return anySize.concat(
-      packSizes.map(packSize => ({
-        label: String(packSize),
-        value: packSize,
-      }))
-    );
-  }, [packSizes]);
+const createPackSizeOption = (line: DraftOutboundLine) => ({
+  packSize: line.packSize,
+  hasAllocated: line.numberOfPacks > 0,
+  hasAvailableStock: isPlaceholder(line)
+    ? true
+    : (line?.stockLine?.availableNumberOfPacks ?? 0) > 0,
+  isPlaceholder: isPlaceholder(line),
+  isExpired: line.expiryDate ? isExpired(new Date(line.expiryDate)) : false,
+  isOnHold: line.stockLine?.onHold,
+  value: line.packSize,
+  label: String(line.packSize),
+});
 
-  return { options, packSizes };
-};
+const createAnyOption = (t: ReturnType<typeof useTranslation>) => () => ({
+  packSize: -1,
+  isAny: true,
+  hasAllocated: false,
+  hasAvailableStock: false,
+  isPlaceholder: false,
+  isExpired: false,
+  isOnHold: undefined,
+  value: -1,
+  label: t('label.any'),
+});
 
 export const usePackSizeController = (lines: DraftOutboundLine[]) => {
-  const { options, packSizes } = usePackSizes(lines);
+  const t = useTranslation('distribution');
 
-  const [selected, setSelected] = useState<
-    | {
-        label: string;
-        value: number;
-      }
-    | undefined
-  >();
+  // The selected pack size for auto allocation. The initial value
+  // will be determined by the lines in the invoice.
+  // A specific pack size is selected if:
+  //     - All lines with [allocated packs] are of [the same] pack size.
+  //     - All lines have [no allocated packs] and are of [the same] pack size.
+  // Any is selected when:
+  //     - All lines with [allocated packs] are of [different] pack sizes
+  //     - All lines have [no allocated packs] and are of [different] pack sizes.
+  const [selected, setSelected] = useState<PackSizeOption | undefined>();
 
-  const items = lines.map(({ item }) => item);
-  const itemId = ifTheSameElseDefault(items, 'id', '');
-  useEffect(() => {
-    setSelected(undefined);
-  }, [itemId]);
+  // Helper to set the pack size with just a number.
+  const setPackSize = (packSize: number) => {
+    const packSizeOption = validPackSizes.find(
+      ({ value }) => value === packSize
+    );
+    if (!packSizeOption) return;
+    setSelected(packSizeOption);
+  };
+
+  // Create the pack size options.
+  const packSizes = lines.map(createPackSizeOption);
 
-  const setPackSize = useCallback(
-    (newValue: number) => {
-      const packSizeOption = options.find(({ value }) => value === newValue);
-      if (!packSizeOption) return;
-      setSelected(packSizeOption);
-    },
-    [options, setSelected]
+  // Valid pack sizes are the pack size of a line which
+  // - is a placeholder and has allocated stock.
+  // - is a placeholder and is the only line.
+  // - No other placeholders.
+  // - is not on hold.
+  // - is not expired.
+  // - has some available stock.
+  const validPackSizes = uniqBy(
```

Comment:
😁  nice - clear and easy to follow, you've made the best of big stack of conditions